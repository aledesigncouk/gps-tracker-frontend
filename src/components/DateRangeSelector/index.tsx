import React, { useState } from "react";
import { getMonth, getYear } from 'date-fns';
import range from "lodash/range";
import DatePicker from "react-datepicker";
import { useStore } from "@store/ContextStore";
import Button from "@components/Button";
import Modal from "@components/Modal";

import style from "@styles/components/DateRangeSelector.module.scss";

const DateRangeSelector = () => {
  const {
    startDate,
    setStartDate,
    endDate,
    setEndDate,
    setRunFetchData,
  } = useStore();
  const [isModalOpen, setIsModalOpen] = useState<boolean>(false);

  const handleFetch = () => {
    setRunFetchData(true);
  };

  const years = range(2021, getYear(new Date()) + 1, 1);
  const months = [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
  ];

  return (
    <>
      <div className={style.datePickerWrapper}>
        <label className={style.label}>Start Date</label>
        <DatePicker
          renderCustomHeader={({
            date,
            changeYear,
            changeMonth,
            decreaseMonth,
            increaseMonth,
            prevMonthButtonDisabled,
            nextMonthButtonDisabled,
          }) => (<div style={{ margin: 10, display: "flex", justifyContent: "center", }}
          >
            <button onClick={decreaseMonth} disabled={prevMonthButtonDisabled}>
              {"<"}
            </button>
            <select
              value={getYear(date)}
              onChange={({ target: { value } }) => changeYear(value as any)}
            >
              {years.map((option) => (
                <option key={option} value={option}>
                  {option}
                </option>
              ))}
            </select>

            <select
              value={months[getMonth(date)]}
              onChange={({ target: { value } }) =>
                changeMonth(months.indexOf(value))
              }
            >
              {months.map((option) => (
                <option key={option} value={option}>
                  {option}
                </option>
              ))}
            </select>

            <button onClick={increaseMonth} disabled={nextMonthButtonDisabled}>
              {">"}
            </button>
          </div>
          )}
          className={style.dateInput}
          customInput={<input data-testid="start-date" type="text" />}
          selected={startDate}
          onChange={(date: Date) => setStartDate(date)}
          selectsStart
          startDate={startDate}
          endDate={endDate}
          dateFormat={"dd / MM / yyyy"}
          withPortal
        />
      </div>

      <div className={style.datePickerWrapper}>
        <label className={style.label}>End Date</label>
        <DatePicker
          renderCustomHeader={({
            date,
            changeYear,
            changeMonth,
            decreaseMonth,
            increaseMonth,
            prevMonthButtonDisabled,
            nextMonthButtonDisabled,
          }) => (<div style={{ margin: 10, display: "flex", justifyContent: "center", }}
          >
            <button onClick={decreaseMonth} disabled={prevMonthButtonDisabled}>
              {"<"}
            </button>
            <select
              value={getYear(date)}
              onChange={({ target: { value } }) => changeYear(value as any)}
            >
              {years.map((option) => (
                <option key={option} value={option}>
                  {option}
                </option>
              ))}
            </select>

            <select
              value={months[getMonth(date)]}
              onChange={({ target: { value } }) =>
                changeMonth(months.indexOf(value))
              }
            >
              {months.map((option) => (
                <option key={option} value={option}>
                  {option}
                </option>
              ))}
            </select>

            <button onClick={increaseMonth} disabled={nextMonthButtonDisabled}>
              {">"}
            </button>
          </div>
          )}
          className={style.dateInput}
          customInput={<input data-testid="end-date" type="text" />}
          selected={endDate}
          onChange={(date: Date) => setEndDate(date)}
          selectsEnd
          startDate={startDate}
          endDate={endDate}
          minDate={startDate}
          dateFormat={"dd / MM / yyyy"}
          withPortal
        />
      </div>
      <Button className={"btn-primary"} onClick={handleFetch} data-testid="fetch-button" label="Fetch Data" />
      <Modal
        data-testid="modal"
        title="Error"
        content="Please select a start and end date."
        isOpen={isModalOpen}
        setModal={setIsModalOpen}
      />
    </>
  );
};

export default DateRangeSelector;
