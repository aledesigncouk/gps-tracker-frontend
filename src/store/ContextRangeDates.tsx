import React, { createContext, useState, ReactNode } from 'react';

interface RangeDatesContextProps {
    startDate: Date | null;
    endDate: Date | null;
    setStartDate: (date: Date | null) => void;
    setEndDate: (date: Date | null) => void;
}

const RangeDatesContext = createContext<RangeDatesContextProps | undefined>(undefined);

export const RangeDatesContextStore: React.FC<{ children: ReactNode }> = ({ children }) => {
    const [startDate, setStartDate] = useState<Date | null>(null);
    const [endDate, setEndDate] = useState<Date | null>(null);

    return (
        <RangeDatesContext.Provider value={{
            startDate: startDate,
            setStartDate: setStartDate,
            endDate: endDate,
            setEndDate: setEndDate,
        }}>
            {children}
        </RangeDatesContext.Provider>
    );
};

export const useRangeDatesStore = () => {
    const context = React.useContext(RangeDatesContext);
    if (context === undefined) {
        throw new Error('useRangeDatesStore must be used within a RangeDatesProvider');
    }
    return context;
};

export default RangeDatesContextStore;