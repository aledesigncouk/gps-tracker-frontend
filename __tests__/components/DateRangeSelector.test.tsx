import { render, screen, fireEvent } from '@testing-library/react';
import DateRangeSelector from '@components/DateRangeSelector';
import { useRangeDatesStore } from '@store/ContextRangeDates';

jest.mock('@store/ContextRangeDates', () => ({
  useRangeDatesStore: jest.fn(),
}));

describe('DateRangeSelector component', () => {
  const setStartDate = jest.fn();
  const setEndDate = jest.fn();

  beforeEach(() => {
    useRangeDatesStore.mockReturnValue({
      startDate: new Date('2023-01-01'),
      endDate: new Date('2023-12-31'),
      setStartDate,
      setEndDate,
    });

    render(<DateRangeSelector />);
  });

  it('renders the DateRangeSelector component', () => {
    const startDateInput = screen.getByTestId('startDate-input');
    const endDateInput = screen.getByTestId('endDate-input');

    expect(startDateInput).toBeInTheDocument();
    expect(endDateInput).toBeInTheDocument();
  });

  it('renders the Start Date label correctly', () => {
    const startDateLabel = screen.getByText('Start Date');
    expect(startDateLabel).toBeInTheDocument();
  });

  it('renders the End Date label correctly', () => {
    const endDateLabel = screen.getByText('End Date');
    expect(endDateLabel).toBeInTheDocument();
  });

  it.skip('fires the setStartDate function when the start date is changed', () => {
    const startDateInput = screen.getByTestId('startDate-input');

    fireEvent.change(startDateInput, { target: { value: '10 / 10 / 2023' } });

    expect(setStartDate).toHaveBeenCalledWith(new Date('2023-10-10'));
  });

  it('fires the setEndDate function when the end date is changed', () => {
    const endDateInput = screen.getByTestId('endDate-input');

    fireEvent.change(endDateInput, { target: { value: '25 / 12 / 2023' } });

    expect(setEndDate).toHaveBeenCalledWith(new Date('2023-12-25'));
  });

  it.skip('disables the End Date input if no Start Date is selected', () => {
    // modal warning test
    useRangeDatesStore.mockReturnValue({
      startDate: null,
      endDate: new Date('2023-12-31'),
      setStartDate,
      setEndDate,
    });

    render(<DateRangeSelector />);

    const endDateInput = screen.getByTestId('endDate-input');
    expect(endDateInput).toBeDisabled();
  });

  it.skip('displays the correct year and month in the custom header', () => {
    const customHeader = screen.getByText('January'); // Start of year date
    expect(customHeader).toBeInTheDocument();
  });

  it.skip('changes the month when clicking the month navigation buttons', () => {
    const prevMonthButton = screen.getByText('<');
    const nextMonthButton = screen.getByText('>');

    fireEvent.click(prevMonthButton);
    fireEvent.click(nextMonthButton);

    expect(prevMonthButton).toBeInTheDocument();
    expect(nextMonthButton).toBeInTheDocument();
  });
});
