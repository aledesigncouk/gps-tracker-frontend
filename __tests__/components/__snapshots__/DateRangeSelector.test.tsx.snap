// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRangeSelector component match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="g-3 row"
  >
    <div
      class="col-sm-5 col-12"
    >
      <label
        class="startDate-label"
      >
        Start Date
      </label>
      <div>
        <div
          class="react-datepicker__input-container"
        >
          <input
            class="startDate-picker-input col-12"
            data-testid="startDate-input"
            type="text"
            value="01 / 01 / 2023"
          />
        </div>
      </div>
    </div>
    <div
      class="col-sm-5 col-12"
    >
      <label
        class="endDate-label"
      >
        End Date
      </label>
      <div>
        <div
          class="react-datepicker__input-container"
        >
          <input
            class="enddate-picker-input col-12"
            data-testid="endDate-input"
            type="text"
            value="31 / 12 / 2023"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
